### PR TITLE
Rename and reorder private functions in `reduction.pxi`

### PR DESCRIPTION
**Pull request by niboshi**
I just want to make `reduction.pxi` a bit more organized and consistent.

This PR only includes renames and reorder of private functions/class.

* `_get_simple_reduction_kernel()` -> `_create_reduction_function()`
* `class simple_reduction_function` -> `class _SimpleReductionKernel`
* `_get_simple_reduction_function()` -> `_SimpleReductionKernel_get_cached_function()`
* `_get_reduction_kernel()` -> `_ReductionKernel_get_cached_function()`

Any comments are welcome (as usual).

### CODE REVIEW

**Comment by asi1024** [comment]
Could you explain why did you reorder the private function declarations? I don't feel the previous version confusing.

File: cupy/core/reduction.pxi [hunk 1]
```diff
@@ -10,7 +10,7 @@ from cupy.cuda import compiler
 from cupy import util
 
 
-cpdef _get_simple_reduction_kernel(
+cpdef _create_reduction_kernel(
```
Comment (asi1024) [inline cupy/core/reduction.pxi]: This function returns `cupy.cuda.function.Function`, not `ReductionKernel`. How about `_create_reduction_function` or `_create_reduction`?
BTW, why not writing the type of return value explicitly for readability?

**Comment by niboshi** [comment]
The first thing is `create_reduction_func()`, this public function is just a wrapper of `class simple_reduction_function`. They should be closely placed.

The next thing is my mistake in #2732 .
`class _AbstractReductionKernel` was placed between `_get_simple_reduction_function()` and `class simple_reduction_function()`, but these two are closely related. `class _AbstractReductionKernel` should not have been placed in the middle of them.

Finally, why the order of these three?
1. `create_reduction_func()`
2. `class simple_reduction_function`
3. `_get_simple_reduction_function()`

That's because it's easier to read (I think) for those who are new to this file.
`create_reduction_func()` is the public entry point. It's natural to begin with it. It uses `class simple_reduction_function` inside it, and naturally the reader can continue reading the class'es implementation. Then it comes to a call to `_get_simple_reduction_function()`. Again naturally the reader can continue reading the implementation of the function.

File: cupy/core/reduction.pxi (hunk 1 above)
Comment (niboshi) [reply cupy/core/reduction.pxi]: `_create_reduction_function` looks good to me.
In that case, I want to rename other functions similarly.
For example,
* `_AbstructReductionKernel._get_kernel()` -> `_get_function()`
* `_ReductionKernel_get_cached_kernel` -> `_Reductionkernel_get_cached_function`
* etc.

Do you think it's ok?

Comment (niboshi) [reply cupy/core/reduction.pxi]: I agree with explicitly typing the return value.

Comment (asi1024) [reply cupy/core/reduction.pxi]: Yes, `_get_function` looks better to me too.

Comment (niboshi) [reply cupy/core/reduction.pxi]: Fixed in fd1d645e91bbf9acad9d92e3b8a6ef2309efa515.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**